From 17ee636b7413c1a45b3a44007861bed98e24f994 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Tue, 14 May 2019 13:47:11 +0200
Subject: [PATCH 141/345] MIPS/OCTEON: Add OCTEON II TLB parity error handling

Add definitions of OCTEON II CvmMemCtl bits and enable TLB parity checks.
Add octeon2_mcheck_handler for TLB parity errors.
  We need to temporarily disable TLB Parity checks early in the handler
  so we don't get into an endless loop.
Add handling of TLB parity errors into EDAC driver.

Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Leonid Rosenboim <lrosenboim@caviumnetworks.com>
Signed-off-by: Carlos Munoz <carlos.munoz@caviumnetworks.com>
---
 arch/mips/cavium-octeon/setup.c       |  4 +++
 arch/mips/include/asm/octeon/octeon.h | 25 ++++++++++++++--
 arch/mips/kernel/genex.S              | 14 ++++++++-
 arch/mips/mm/c-octeon.c               | 41 +++++++++++++++++++++++++++
 drivers/edac/octeon_edac-pc.c         |  8 +++++-
 5 files changed, 87 insertions(+), 5 deletions(-)

diff --git a/arch/mips/cavium-octeon/setup.c b/arch/mips/cavium-octeon/setup.c
index 0cadea8fa154..28e830cfff97 100644
--- a/arch/mips/cavium-octeon/setup.c
+++ b/arch/mips/cavium-octeon/setup.c
@@ -519,6 +519,10 @@ void octeon_user_io_init(void)
 	 * mode. */
 	cvmmemctl.s.cvmsegenau = 0;
 
+	/* Enable TLB parity error reporting on OCTEON II */
+	if (OCTEON_IS_OCTEON2())
+		cvmmemctl.s.tlbperrena = 1;
+
 	write_c0_cvmmemctl(cvmmemctl.u64);
 
 	/* Setup of CVMSEG is done in kernel-entry-init.h */
diff --git a/arch/mips/include/asm/octeon/octeon.h b/arch/mips/include/asm/octeon/octeon.h
index a1cdf82ffb01..aa6047e6aced 100644
--- a/arch/mips/include/asm/octeon/octeon.h
+++ b/arch/mips/include/asm/octeon/octeon.h
@@ -182,8 +182,26 @@ union octeon_cvmemctl {
 		__BITFIELD_FIELD(uint64_t ptgbist:1,
 		/* RO 1 = BIST fail, 0 = BIST pass */
 		__BITFIELD_FIELD(uint64_t wbfbist:1,
-		/* Reserved */
-		__BITFIELD_FIELD(uint64_t reserved:17,
+		__BITFIELD_FIELD(uint64_t utlbbist:1,
+		__BITFIELD_FIELD(uint64_t dutbist:1,
+		__BITFIELD_FIELD(uint64_t disncbpref:1,
+		__BITFIELD_FIELD(uint64_t wbfltime2:1,
+		__BITFIELD_FIELD(uint64_t wbthresh2:1,
+		__BITFIELD_FIELD(uint64_t cvmsegiodis:1,
+		__BITFIELD_FIELD(uint64_t lmtena:1,
+		__BITFIELD_FIELD(uint64_t lmtline:6,
+		/* When set, TLB parity errors can occur. */
+		__BITFIELD_FIELD(uint64_t tlbperrena:1,
+		/* OCTEON II - When set, CVMSET LM parity errors are enabled. */
+		__BITFIELD_FIELD(uint64_t lmemperrena:1,
+		/* OCTEON II - If set, NUDGE/WRITEBACK_INVALIDATE,
+		 * NUDGE_WB, EVICT_SOON, LC, CONT_WRITE_BACK,
+		 * PREPARE_FOR_STORE and PREPARE_FOR_STORE_THROUGH
+		 * prefetch operations become NOPs. */
+		__BITFIELD_FIELD(uint64_t disstpref:1,
+		/* OCTEON II - If set, NORMAL and NOTL2 prefetch
+		 * operations become NOPs. */
+		__BITFIELD_FIELD(uint64_t disldpref:1,
 		/* OCTEON II - TLB replacement policy: 0 = bitmask LRU; 1 = NLU.
 		 * This field selects between the TLB replacement policies:
 		 * bitmask LRU or NLU. Bitmask LRU maintains a mask of
@@ -279,7 +297,7 @@ union octeon_cvmemctl {
 		/* R/W Size of local memory in cache blocks, 54 (6912
 		 * bytes) is max legal value. */
 		__BITFIELD_FIELD(uint64_t lmemsz:6,
-		;)))))))))))))))))))))))))))))))))
+		;))))))))))))))))))))))))))))))))))))))))))))
 	} s;
 };
 
@@ -423,6 +441,7 @@ int unregister_co_cache_error_notifier(struct notifier_block *nb);
 #define CO_CACHE_ERROR_RECOVERABLE 0
 #define CO_CACHE_ERROR_UNRECOVERABLE 1
 #define CO_CACHE_ERROR_WB_PARITY 2
+#define CO_CACHE_ERROR_TLB_PARITY 3
 
 extern unsigned long long cache_err_dcache[NR_CPUS];
 
diff --git a/arch/mips/kernel/genex.S b/arch/mips/kernel/genex.S
index 0594dc9f6ff5..9bab65bba67c 100644
--- a/arch/mips/kernel/genex.S
+++ b/arch/mips/kernel/genex.S
@@ -433,6 +433,18 @@ NESTED(nmi_handler, PT_SIZE, sp)
 	TRACE_IRQS_OFF
 	.endm
 
+	.macro	__build_clear_mce
+#ifdef CONFIG_CPU_CAVIUM_OCTEON
+	/* Disable TLB Parity checks. */
+	dmfc0	t0, $11, 7
+	dli	t1, (1 << 44)
+	or	t0, t0, t1
+	xor	t0, t0, t1
+	dmtc0	t0, $11, 7
+#endif
+	__build_clear_cli
+	.endm
+
 	.macro	__build_clear_fpe
 	.set	push
 	/* gas fails to assemble cfc1 for some archs (octeon).*/ \
@@ -523,7 +535,7 @@ NESTED(nmi_handler, PT_SIZE, sp)
 #else
 	BUILD_HANDLER watch watch sti verbose		/* #23 */
 #endif
-	BUILD_HANDLER mcheck mcheck cli verbose		/* #24 */
+	BUILD_HANDLER mcheck mcheck mce verbose		/* #24 */
 	BUILD_HANDLER mt mt sti silent			/* #25 */
 	BUILD_HANDLER dsp dsp sti silent		/* #26 */
 	BUILD_HANDLER reserved reserved sti verbose	/* others */
diff --git a/arch/mips/mm/c-octeon.c b/arch/mips/mm/c-octeon.c
index 0ee9240f8588..ee8686575c09 100644
--- a/arch/mips/mm/c-octeon.c
+++ b/arch/mips/mm/c-octeon.c
@@ -211,6 +211,46 @@ static int octeon2_be_handler(struct pt_regs *regs, int is_fixup)
 		return MIPS_BE_FATAL;
 }
 
+/*
+ * Octeon specific MachineCheck handler, as TLB parity errors
+ * trigger MachineCheck errors.
+ */
+static int octeon2_mcheck_handler(struct pt_regs *regs)
+{
+	u64 dcache_err;
+	u64 tlbperr_mask = 1ULL << 5;
+	dcache_err = read_octeon_c0_dcacheerr();
+	if (dcache_err & tlbperr_mask) {
+		int rv;
+		union octeon_cvmemctl cvmmemctl;
+
+		/* Clear the indicator */
+		write_octeon_c0_dcacheerr(tlbperr_mask);
+		/*
+		 * Blow everything away to (hopefully) write good
+		 * parity to all TLB entries
+		 */
+		local_flush_tlb_all();
+		/* Reenable TLB parity error reporting. */
+		cvmmemctl.u64 = read_c0_cvmmemctl();
+		cvmmemctl.s.tlbperrena = 1;
+		write_c0_cvmmemctl(cvmmemctl.u64);
+
+		rv = raw_notifier_call_chain(&co_cache_error_chain,
+					     CO_CACHE_ERROR_TLB_PARITY,
+					     NULL);
+		if ((rv & ~NOTIFY_STOP_MASK) != NOTIFY_OK) {
+			unsigned int coreid = cvmx_get_core_num();
+
+			pr_err("Core%u: TLB parity error:\n", coreid);
+			return MIPS_MC_FATAL;
+		}
+
+		return MIPS_MC_DISCARD;
+	}
+	return MIPS_MC_NOT_HANDLED;
+}
+
 /**
  * Probe Octeon's caches
  *
@@ -260,6 +300,7 @@ static void probe_octeon(void)
 		c->options |= MIPS_CPU_PREFETCH;
 
 		board_be_handler = octeon2_be_handler;
+		board_mcheck_handler = octeon2_mcheck_handler;
 		break;
 
 	case CPU_CAVIUM_OCTEON3:
diff --git a/drivers/edac/octeon_edac-pc.c b/drivers/edac/octeon_edac-pc.c
index 73fc27a2711a..392709f3208e 100644
--- a/drivers/edac/octeon_edac-pc.c
+++ b/drivers/edac/octeon_edac-pc.c
@@ -55,6 +55,12 @@ static int  co_cache_error_event(struct notifier_block *this,
 				   core, cpu);
 		edac_device_handle_ue(p->ed, cpu, 2, "write-buffer");
 		return NOTIFY_STOP;
+	case CO_CACHE_ERROR_TLB_PARITY:
+		edac_device_printk(p->ed, KERN_ERR,
+				   "TLB parity error: core %d/cpu %d\n",
+				   core, cpu);
+		edac_device_handle_ue(p->ed, cpu, 3, "TLB-parity");
+		return NOTIFY_STOP;
 	default:
 		WARN(1, "Unknown event: %lu\n", event);
 		return NOTIFY_BAD;
@@ -99,7 +105,7 @@ static int co_cache_error_probe(struct platform_device *pdev)
 	platform_set_drvdata(pdev, p);
 
 	p->ed = edac_device_alloc_ctl_info(0, "cpu", num_possible_cpus(),
-					   "cache", 3, 0, NULL, 0,
+					   "cache", 4, 0, NULL, 0,
 					   edac_device_alloc_index());
 	if (!p->ed)
 		goto err;
-- 
2.25.1

